lib/connect.js

```diff
@@ -234,29 +244,39 @@ const getTasks = (model, setupListeners) => {
         setupListeners(mongoClient);
       }
 
-      mongoClient.connect((err, _client) => {
-        ctx(err);
-
-        if (err) {
-          if (tunnel) {
-            debug('data-service connection error, shutting down ssh tunnel');
-            tunnel.close();
-          }
-
-          return cb(err);
+      /** @type {Promise<never>} */
+      const waitForTunnelError = new Promise(async(_, reject) => {
+        if (tunnel) {
+          const [error] = await once(tunnel, 'error');
+          reject(error);
         }
+      });
 
-        client = _client;
-
+      const closeTunnelOnError = async() => {
         if (tunnel) {
-          client.on('close', () => {
-            debug('data-service disconnected. shutting down ssh tunnel');
-            tunnel.close();
-          });
+          debug('data-service connection error, shutting down ssh tunnel');
+          try {
+            await tunnel.close();
+            debug('ssh tunnel stopped');
+          } catch (err) {
+            debug('ssh tunnel stopped with error: %s', err.message);
+          }
         }
+      };
 
-        cb(null, { url: model.driverUrlWithSsh, options: validOptions });
-      });
+      try {
+        const _client = await Promise.race([
+          mongoClient.connect(),
+          waitForTunnelError
+        ]);
```

Comment:
Not really happy with this bit here as we are not really waiting for `MongoClient.connect` to resolve if tunnel threw, but that's more or less the only way to do it without doing weird callback stuff as it was in the previous implementation

Reply:
Good point!

Are user actionable errors about the tunnel thrown by `.listen`?

The one thing I was assuming here is this: all the ssh related errors are thrown on `tunnel.listen` ie if a cert or password is invalid .... and the only issues catched by on('error') are not  caused by user mistakes (ie. they picked a wrong file) but by programming bug or network conditions that are not actionable beyond a retry: exception when the tcp connection is opened by MongoClient. 

If that is true i think we can even ignore the race with `waitForTunnelError`, since MongoClient will throw a connection error and we could just give a nice error in the UI: `The tunnel was open correctly but the connection to mongodb failed: ${mongoError}`.

For sure we could give more info but is super fine, we are making it connect when it should, which now does not, we already improve the user life immensely.

If however instead errors like wrong password can only be catched by `waitForTunnelError` than is a different story and probably we should wait for that as you did, or better make sure those errors are thrown by listen.

Reply:
> all the ssh related errors are thrown on tunnel.listen ie if a cert or password is invalid

This doesn't happen until first connection happens, tunnel is established for every incoming connection to ssh tunnel server, so things like forward out are just not happening before that. If we want them to be thrown when `listen` is called, some kind of ping request needs to be issued, which is doable. What do you think, should we add something like this here?

So in current implementation both errors emitted by tunnel or provided in callback can be helpful. But as mentioned, if tunnel emitted one, it probably will have a higher priority in terms of being useful to the user.

Reply:
> forward out are just not happening before that

When it gets to forward the connection to ssh should already be established, isn't it?

> tunnel is established for every incoming connection to ssh

why? can't we just both create a new tunnel and 'forwarding tcp server' on listen instead?

Reply:
> why? can't we just both create a new tunnel and 'forwarding tcp server' on listen instead?

`ssh2` docs don't recommend reusing one client for multiple connections, so @mongodb-js/ssh-tunnel [creates new client for every incoming connection](https://github.com/mongodb-js/ssh-tunnel/blob/907b7627a7d04b5bc651db3a71b135f6bd012976/src/index.ts#L73), which also means that client is not created until incoming connection is received